src/jit/gentree.cpp

```diff
@@ -11926,16 +11966,27 @@ GenTree* Compiler::gtFoldTypeCompare(GenTree* tree)
         // We can't answer the equality comparison definitively at jit
         // time, but can still simplfy the comparison.
         //
+        // Find out how we can compare the two handles.
+        // NOTE: We're potentially passing NO_CLASS_HANDLE, but the runtime knows what to do with it here.
+        CorInfoInlineTypeCheck inliningKind = info.compCompHnd->canInlineTypeCheck(cls1Hnd, CORINFO_INLINE_TYPECHECK_SOURCE_TOKEN);
+        assert(inliningKind == CORINFO_INLINE_TYPECHECK_PASS || inliningKind == CORINFO_INLINE_TYPECHECK_USE_HELPER);
+
+        // If the first type needs helper, we're done and use helper. Otherwise check the other type.
```

Comment:
This is missing some cases. For example, consider `cls1Hnd` is `__Canon` and `cls2Hnd` is a type that is guaranteed to have unique EEType. We can still compare pointers directly in this case. 

I am wondering whether this should be: `if (inliningKind == CORINFO_INLINE_TYPECHECK_USE_HELPER)`. The idea is to try to get a better answer by looking at the other side if we got a helper from one side.

Reply:
Good catch! Fixed!